daffodil-io/src/main/scala/org/apache/daffodil/io/InputSourceDataInputStream.scala

```diff
@@ -117,6 +118,57 @@ final class InputSourceDataInputStream private (val inputSource: InputSource)
   @inline
   override final def bitLimit0b: MaybeULong = cst.bitLimit0b
 
+  /**
+   * Tells us if the underlying input source has detected end-of-data
+   * (the read(...) call returned -1.
+   *
+   * But this does NOT tell us we are positioned at the end, only whether
+   * in the course of reading, we encountered the end of data. If we
+   * backtracked we could have seen the end of data, but backed up in
+   * the data to an earlier position.
+   */
+  def hasReachedEndOfData: Boolean = inputSource.hasReachedEndOfData
+
+  /**
+   * Determine if we're positioned at the end of data without
+   * doing any additional blocking operation such as reading more
+   * data to test if there is any.
+   *
+   * This depends on the underlying inputSource keeping track of
+   * whether it has previously hit the end of data or not.
+   *
+   * @return
+   */
+  final def isAtEnd(): Boolean = {
+    if (bitLimit0b.isDefined) {
+      bitPos0b == bitLimit0b.get
+    } else {
+      // There is no bit limit, so whether we're at the end
+      // requires that we've hit then end-of-data on a read
+      // operation, AND that we've not backtracked since then,
+      // meaning we're positioned at the end of the last byte
+      // read.
+      hasReachedEndOfData && {
```

Comment:
I think this is a now a non-backwards compatible change now. isAtEnd can only be true if we've gotten an EOD. And I imagine in a lot of binary formats we'll never try that extra read to get the EOD because we don't ever need to scan.